futures-util/src/task/local_waker_ref.rs

```diff
@@ -0,0 +1,140 @@
+#![allow(cast_ptr_alignment)] // clippy is too strict here
+
+use std::marker::PhantomData;
+use std::ops::Deref;
+use std::ptr::NonNull;
+use std::sync::Arc;
+use std::task::{LocalWaker, Waker, Wake, UnsafeWake};
+
+/// A [`LocalWaker`](::std::task::LocalWaker) that is only valid for a given lifetime.
+///
+/// Note: this type implements [`Deref<Target = LocalWaker>`](::std::ops::Deref),
+/// so it can be used to get a `&LocalWaker`.
+#[derive(Debug)]
+pub struct LocalWakerRef<'a> {
+    local_waker: LocalWaker,
+    _marker: PhantomData<&'a ()>,
+}
+
+impl<'a> LocalWakerRef<'a> {
+    /// Create a new [`LocalWakerRef`] from a [`LocalWaker`].
+    ///
+    /// Note: this function is safe, but it is generally only used
+    /// from `unsafe` contexts that need to create a `LocalWaker`
+    /// that is guaranteed not to outlive a particular lifetime.
+    pub fn new(local_waker: LocalWaker) -> Self {
+        LocalWakerRef {
+            local_waker,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<'a> Deref for LocalWakerRef<'a> {
+    type Target = LocalWaker;
+
+    fn deref(&self) -> &LocalWaker {
+        &self.local_waker
+    }
+}
+
+// Pointers to this type below are really pointers to `Arc<W>`
+struct ReferencedArc<W> {
+    _marker: PhantomData<W>,
+}
+
+unsafe impl<W: Wake + 'static> UnsafeWake for ReferencedArc<W> {
+    #[inline]
+    unsafe fn clone_raw(&self) -> Waker {
+        let me = self as *const ReferencedArc<W> as *const Arc<W>;
+        Arc::clone(&*me).into()
+    }
+
+    #[inline]
+    unsafe fn drop_raw(&self) {}
+
+    #[inline]
+    unsafe fn wake(&self) {
+        let me = self as *const ReferencedArc<W> as *const Arc<W>;
+        W::wake(&*me)
+    }
+
+    #[inline]
+    unsafe fn wake_local(&self) {
+        let me = self as *const ReferencedArc<W> as *const Arc<W>;
+        W::wake_local(&*me)
```

Comment:
and here

Reply:
and on the other type

Reply:
basically like in `futures_unordered/task.rs`

https://github.com/rust-lang-nursery/futures-rs/blob/c8e7446b8d31c3ae0fad6f5d98a830db770fc86a/futures-util/src/stream/futures_unordered/task.rs#L207-L212